### ISSUE TITLE
Audit Log Part E (response body) is logged for no obvious reason

### ISSUE THREAD

**Issue by studersi** [coreruleset/coreruleset#2530]
### Describe the bug

The audit log part E (response body) is often logged where there is no obvious reason it should be logged. For example, on a service where audit log part E is disabled, a request exceeding PCRE limits would log the E part despite no rule being triggered adding this part to the audit log.

### Steps to reproduce

1. Configure SecAuditLogParts to not use log part E.
2. Execute request that ends up in audit log without triggering a rule explicitly adding the audit log part E using `ctl:auditLogParts=+E`
3. The audit log will contain the audit log part E anyway.

### Expected behaviour

The audit log should not contain the audit log part E for the request described above.

### Actual behaviour

The audit log does contain the audit log part E for the request described above.

### Additional context

There are chained rules in CRS where the audit log part E is added because it is relevant to the rule that is triggered: `ctl:auditLogParts=+E`.

Some of these rules are chained rules. For chained rules the audit log part is added in the base rule, not the chained rule. For example:
```
SecRule ARGS "@rx ^(?i:file|ftps?|https?)://(?:[^@]+@)?([^/]+)" \
    "id:931130,\
    phase:2,\
    block,\
    capture,\
    t:none,\
    msg:'Possible Remote File Inclusion (RFI) Attack: Off-Domain Reference/Link',\
    logdata:'Matched Data: %{TX.0} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',\
    tag:'application-multi',\
    tag:'language-multi',\
    tag:'platform-multi',\
    tag:'attack-rfi',\
    tag:'OWASP_CRS',\
    tag:'capec/1000/152/175/253',\
    tag:'paranoia-level/2',\
    ctl:auditLogParts=+E,\
    ver:'OWASP_CRS/4.0.0-rc1',\
    severity:'CRITICAL',\
    setvar:'tx.rfi_parameter_%{MATCHED_VAR_NAME}=.%{tx.1}',\
    chain"
    SecRule TX:/rfi_parameter_.*/ "!@endsWith .%{request_headers.host}" \
        "setvar:'tx.rfi_score=+%{tx.critical_anomaly_score}',\
        setvar:'tx.inbound_anomaly_score_pl2=+%{tx.critical_anomaly_score}'"
```

According to the documentation for chaining rules ([chain](https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-(v2.x)#chain)), some non-disruptive actions will be executed on non-disruptive rules (most CRS rules are non-disruptive) even if the chained rules are not satisfied:
> Non-disruptive rules can be used in any rule; they will be executed if the rule that contains them matches and not only when the entire chain matches. 

Indeded, if `ctl:auditLogParts=+E` is moved to the chained rule, the problem disappears. For example:
```
SecRule ARGS "@rx ^(?i:file|ftps?|https?)://(?:[^@]+@)?([^/]+)" \
    "id:931130,\
    phase:2,\
    block,\
    capture,\
    t:none,\
    msg:'Possible Remote File Inclusion (RFI) Attack: Off-Domain Reference/Link',\
    logdata:'Matched Data: %{TX.0} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',\
    tag:'application-multi',\
    tag:'language-multi',\
    tag:'platform-multi',\
    tag:'attack-rfi',\
    tag:'OWASP_CRS',\
    tag:'capec/1000/152/175/253',\
    tag:'paranoia-level/2',\
    ver:'OWASP_CRS/4.0.0-rc1',\
    severity:'CRITICAL',\
    setvar:'tx.rfi_parameter_%{MATCHED_VAR_NAME}=.%{tx.1}',\
    chain"
    SecRule TX:/rfi_parameter_.*/ "!@endsWith .%{request_headers.host}" \
        "ctl:auditLogParts=+E,\
        setvar:'tx.rfi_score=+%{tx.critical_anomaly_score}',\
        setvar:'tx.inbound_anomaly_score_pl2=+%{tx.critical_anomaly_score}'"
```

**Comment by studersi** [comment]
Added a PR that should fix this bug: https://github.com/coreruleset/coreruleset/pull/2531.

**Comment by airween** [comment]
Hi @studersi,

thanks for report - and for the PR too :).

We will discuss your fix soon - thanks again.

**Comment by fzipi** [comment]
@airween I think this is ready to merge. The description is clear, and the PR works as intended. Follow-up work might be to add a check using `msc-pyparser`so this doesn't happen again in future rules? And maybe adding a comment on the CONTRIBUTING.md?

**Comment by airween** [comment]
> Follow-up work might be to add a check using `msc-pyparser`so this doesn't happen again in future rules? And maybe adding a comment on the CONTRIBUTING.md?

`msc-pyparser` is not able to do this, but I think I understand what you mean. We have a [check script](https://github.com/coreruleset/coreruleset/tree/v4.0/dev/util/crs-rules-check), I suppose we should add there a new test method.

Not is has 5 steps:
* [syntax check](https://github.com/coreruleset/coreruleset/tree/v4.0/dev/util/crs-rules-check#test-1---syntax-check)
* [case sensitive test](https://github.com/coreruleset/coreruleset/tree/v4.0/dev/util/crs-rules-check#test-2---case-sensitive-test)
* [wrong action ordering](https://github.com/coreruleset/coreruleset/tree/v4.0/dev/util/crs-rules-check#test-3---wrong-action-ordering)
* [wrong indentation](https://github.com/coreruleset/coreruleset/tree/v4.0/dev/util/crs-rules-check#test-4---wrong-indentation)
* [empty implicit `@rx` operator](https://github.com/coreruleset/coreruleset/tree/v4.0/dev/util/crs-rules-check#test-5---empty-implicit-rx-operator)

So if I understand correctly, we need a test case which checks that `ctl:auditLogParts=+E` action can occur only in last rule in every chained rule. Is there any criteria? I mean any `ctl` action can occur only in last part...?

**Comment by fzipi** [comment]
Probably yes. But the tricky part is that we might want things in intermediate steps? 🤔 The case for logging is clear, but not others I guess....

**Comment by fzipi** [comment]
Thanks for reminding me. I think we are good! Closing.
